src/v/storage/segment.cc

```diff
@@ -96,6 +96,8 @@ ss::future<> segment::close() {
     co_await do_flush();
     co_await do_close();
 
+    vassert(_inflight.empty(), "Closing segment with inflight writes.");
+
```

Comment:
>because we use the segment stable offset as reader upper
>bound, this can lead to a situation where acked data is not visible and
>worse consequences.

Just making sure, even though we're adding to `_inflight` before appending, we aren't at risk for having readers read that data though right (especially important since the append may not have been dispatched)? It's just that now, `advance_stable_offset` will at least be able to consider the in-flight write even if we haven't called append() yet?